### ISSUE TITLE
Add WCAG accessibility support to web components

### ISSUE THREAD

**Issue by nicolatamburini** [magma-design-system/magma#125]
***In GitLab by @vittorio.vittori on Sep 18, 2022, 10:42 GMT+2:***

We should deploy a suite of web components that simplify WCAG compatibility where possible in order to be compliant with key accessibility standards for our users.

- [The Guide to Accessible Web Components](https://www.erikkroes.nl/blog/accessibility/the-guide-to-accessible-web-components-draft/#accessible-ui-components)
- [How to Meet WCAG](https://www.w3.org/WAI/WCAG21/quickref/?showtechniques=111)
- [Introduzione per tutti all’accessibilità web](https://www.youtube.com/watch?v=B0WdYUNDgx4)

## Proposal impact

Impacts the following Design System projects:

- Stencil (web components)

Impacts the following Design System environment/tool:

- CI (Continuous Integration)

Impacts **ALL** Maggioli products that uses web components.

## Related problem

The main problem is that currently our web components do not give elements of accessibility for those suffering from cognitive disabilities, we should give an order to the priorities and levels of accessibility and integrate them into our suite of components.

## The solution i'd like to make

To make the right solution it is necessary to carefully evaluate the type of intervention and gradually integrate all the necessary features.

## Alternatives i've considered

We should also consider the presence of dedicated linters so as not to have to repeat these steps for future components as well.

*Migrated from GitLab: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/magma/design-system/-/work_items/125*

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Sep 18, 2022, 14:15 GMT+2:***

Perhaps it is better to open an issue for each web component with a to do list of the accessibility features to be integrated (if applicable), for example:

Component name
- ARIA accessibility standard (list of things to check)
- WCAG standards (list of things to check)
- Automatic focus and keyboard support

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Sep 23, 2022, 22:10 GMT+2:***

created branch [`125-add-wcag-accessibility-support-to-web-components`](/maggiolispa/ricerca-sviluppo-new-media/design-system/-/compare/dev...125-add-wcag-accessibility-support-to-web-components) to address this issue

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Sep 26, 2022, 10:05 GMT+2:***





## Component infos



| Component name | Infos |
|-|-|
| `mds-avatar` | [Component URL](https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/stencil/src/components/mds-accordion) |

## WCAG Checklist

Guidance for writing, designing, and developing for accessibility, [these tips](https://www.w3.org/WAI/tips/) introduce some basic considerations for making your website more accessible to people with disabilities.

Links provide links to additional guidance.

#### 1. Writing



- [ ] [Provide informative, unique page titles](https://www.w3.org/WAI/tips/writing/#provide-informative-unique-page-titles)
- [ ] [Use headings to convey meaning and structure](https://www.w3.org/WAI/tips/writing/#use-headings-to-convey-meaning-and-structure)
- [ ] [Make link text meaningful](https://www.w3.org/WAI/tips/writing/#make-link-text-meaningful)
- [ ] [Write meaningful text alternatives for images](https://www.w3.org/WAI/tips/writing/#write-meaningful-text-alternatives-for-images)
- [ ] [Create transcripts and captions for multimedia](https://www.w3.org/WAI/tips/writing/#create-transcripts-and-captions-for-multimedia)
- [ ] [Provide clear instructions](https://www.w3.org/WAI/tips/writing/#provide-clear-instructions)
- [ ] [Keep content clear and concise](https://www.w3.org/WAI/tips/writing/#keep-content-clear-and-concise)

#### 2. Designing



- [ ] [Provide sufficient contrast between foreground and background](https://www.w3.org/WAI/tips/designing/#provide-sufficient-contrast-between-foreground-and-background)
- [ ] [Don’t use color alone to convey information](https://www.w3.org/WAI/tips/designing/#dont-use-color-alone-to-convey-information)
- [ ] [Ensure that interactive elements are easy to identify](https://www.w3.org/WAI/tips/designing/#ensure-that-interactive-elements-are-easy-to-identify)
- [ ] [Provide clear and consistent navigation options](https://www.w3.org/WAI/tips/designing/#provide-clear-and-consistent-navigation-options)
- [ ] [Ensure that form elements include clearly associated labels](https://www.w3.org/WAI/tips/designing/#ensure-that-form-elements-include-clearly-associated-labels)
- [ ] [Provide easily identifiable feedback](https://www.w3.org/WAI/tips/designing/#provide-easily-identifiable-feedback)
- [ ] [Use headings and spacing to group related content](https://www.w3.org/WAI/tips/designing/#use-headings-and-spacing-to-group-related-content)
- [ ] [Create designs for different viewport sizes](https://www.w3.org/WAI/tips/designing/#create-designs-for-different-viewport-sizes)
- [ ] [Include image and media alternatives in your design](https://www.w3.org/WAI/tips/designing/#include-image-and-media-alternatives-in-your-design)
- [ ] [Provide controls for content that starts automatically](https://www.w3.org/WAI/tips/designing/#provide-controls-for-content-that-starts-automatically)

#### 3. Developing



- [ ] [Associate a label with every form control](https://www.w3.org/WAI/tips/designing/#associate-a-label-with-every-form-control)
- [ ] [Include alternative text for images](https://www.w3.org/WAI/tips/designing/#include-alternative-text-for-images)
- [ ] [Identify page language and language changes](https://www.w3.org/WAI/tips/designing/#identify-page-language-and-language-changes)
- [ ] [Use mark-up to convey meaning and structure](https://www.w3.org/WAI/tips/designing/#use-mark-up-to-convey-meaning-and-structure)
- [ ] [Help users avoid and correct mistakes](https://www.w3.org/WAI/tips/designing/#help-users-avoid-and-correct-mistakes)
- [ ] [Reflect the reading order in the code order](https://www.w3.org/WAI/tips/designing/#reflect-the-reading-order-in-the-code-order)
- [ ] [Write code that adapts to the user’s technology](https://www.w3.org/WAI/tips/designing/#write-code-that-adapts-to-the-users-technology)
- [ ] [Provide meaning for non-standard interactive elements](https://www.w3.org/WAI/tips/designing/#provide-meaning-for-non-standard-interactive-elements)
- [ ] [Ensure that all interactive elements are keyboard accessible](https://www.w3.org/WAI/tips/designing/#ensure-that-all-interactive-elements-are-keyboard-accessible)
- [~] [Avoid CAPTCHA where possible](https://www.w3.org/WAI/tips/designing/#avoid-captcha-where-possible)

#### 4. ARIA support

Remember, [no ARIA is better than bad ARIA](https://www.w3.org/WAI/ARIA/apg/practices/read-me-first/), if you need to use ARIA, make sure you understand the [ARIA roles, states, and properties](https://www.w3.org/TR/wai-aria-1.1/#roles_states_props) and [ARIA design patterns](https://www.w3.org/TR/wai-aria-practices-1.1/).

Find the best [pattern](https://www.w3.org/WAI/ARIA/apg/patterns/) for the component you are building, and make sure you understand the [ARIA roles, states, and properties](https://www.w3.org/TR/wai-aria-1.1/#roles_states_props) that are used in the pattern.

---

## Web Accessibility evaluation tools list

You can find the [Web Accessibility evaluation tools list here][wael], we reported it below.

- [WCAG 2.1 Scan Tool][wcag-scan]: Get a free WCAG 2.1 A/AA compliance summary of your website from our A11Y Compliance Platform.
- [Web Accessibility][wa]: Accessibility tools for your entire development lifecycle. Need to test an app or secure site? We have tools for that. Our tools provide additional flexibility for testing sites requiring authentication and mobile apps. Download our browser extension or development tools to start your accessibility journey.
- [A11Y Color Contrast Accessibility Validator][a11y-color]: A free website compliance tool that displays the color contrast issues of a web page per WCAG Guidelines. The results display color combinations that fail the contrast checkpoints and provide specific recommendations on how to fix the issue to become compliant.
- [WAVE]: A suite of evaluation tools that helps authors make their web content more accessible to individuals with disabilities. WAVE can identify many accessibility and Web Content Accessibility Guideline (WCAG) errors, but also facilitates human evaluation of web content.

[wael]: https://www.w3.org/WAI/ER/tools/
[wave]: https://wave.webaim.org/
[wa]: https://www.webaccessibility.com/tools/
[a11y-color]: https://color.a11y.com/?wc3
[wcag-scan]: https://www.boia.org/w3c-tools-services-a11y




[ci]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/pipelines
[design-tokens]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/design-tokens
[figma-web-components]: https://www.figma.com/file/evxGyMuj9309w6BZXXz4Ls/Magma-[Palette dei colori compatibile con quella di Synbee](https://github.com/magma-design-system/magma/milestone/2)F-Web-Component
[figma]: https://www.figma.com/files/team/945687796790515562
[gitlab]: https://gitlab.com/
[icons]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/icons
[identity]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/identity
[react]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/react
[stencil]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/stencil
[styles]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/styles
[svg-icons]: https://gitlab.com/maggiolispa/ricerca-sviluppo-new-media/design-system/-/tree/dev/projects/svg-icons

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Sep 26, 2022, 10:06 GMT+2:***

I have added a sample template, please give me feedback if that's ok for you.

**Comment by nicolatamburini** [comment]
***In GitLab by @andrea.pruccoli on Sep 26, 2022, 10:19 GMT+2:***

I appreciate the presence of so many links pointing to the documentation for each topic, it will help a lot every time, at least for me will be useful for checking the meaning of every point :smile: 

It could be a bit of `wall-text`, but I think there are more pros than cos doing so.

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Sep 26, 2022, 15:20 GMT+2:***

yep, in the end I decided to insert so much material because actually the interventions on the single components are quite deep, I think it is necessary to know all this information in order to correctly integrate the accessibility features.

**Comment by nicolatamburini** [comment]
***In GitLab by @vittorio.vittori on Oct 4, 2022, 11:11 GMT+2:***

mentioned in commit 28a769f710462778dfea6b88f653b09d2df42056
